### PR TITLE
Add numberOfVectors to get_stats api

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add numberOfVectors to get_stats api

* **What is the current behavior?** (You can also link to an open issue here)
we only have numberOfDocuments

* **What is the new behavior (if this is a feature change)?**
we now have numberOfVectors attribute

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no.

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
running

* **Related Python client changes** (link commit/PR here)
working on

* **Related documentation changes** (link commit/PR here)
working on 

* **Other information**:
no

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/tensor_search/tensor_search.py
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py:1]: We should do some error handling on the requests to Marqo-os.

For example, if we get a non-200 response, we could have something like this: 
```
"numberOfVectors": null
"errors": [
    "error calculating numerOfVectors: error response from Marqo-os /_stats endpoint: <error message>"
]
```
To be honest, I am not sure what the best way to communicate these errors is. It is best to see what other APIs do. Also maybe @farshidz has some ideas

File: tests/tensor_search/test_get_stats.py
Comment (pandu-k) [inline tests/tensor_search/test_get_stats.py:1]: Let's get some edge case and HTTP error tests

File: src/marqo/tensor_search/tensor_search.py [hunk 1]
```diff
@@ -274,17 +274,38 @@ def get_stats(config: Config, index_name: str):
...
+        }
+    }
+
     try:
         doc_count = HttpRequests(config).post(path=F"{index_name}/_count")["count"]
-        nested_doc_count = HttpRequests(config).get(path=F"{index_name}/_stats/docs")["indices"][index_name]["primaries"][
-            "docs"]["count"]
+        vector_count = HttpRequests(config).get(path=f"{index_name}/_search", body=body) \
+            ["aggregations"]["nested_chunks"]["chunk_count"]["value"]
     except KeyError as e:
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py]: nice job catching KeyError.

You may also need to catch TypeError, if you are parsing an unknown dict

File: tests/tensor_search/test_get_stats.py
Comment (pandu-k) [inline tests/tensor_search/test_get_stats.py:1]: Let's also get some cases where the HTTP response isn't expected (like an unknown dict type because of an error)

File: src/marqo/tensor_search/tensor_search.py (hunk 1 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py]: thanks! added!

File: src/marqo/tensor_search/tensor_search.py [hunk 2]
```diff
@@ -267,9 +267,53 @@ def _autofill_index_settings(index_settings: dict):
 
 
 def get_stats(config: Config, index_name: str):
-    doc_count = HttpRequests(config).post(path=F"{index_name}/_count")["count"]
+    """Returns the number of documents and vectors in the index.
+
+    The _count API counts top-level documents.
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: I suggest moving these details out of method docstring and into a standard comment. The docstring is generally a blackbox description of what the method does and details that are relevant for usage of the method (e.g., side effects, time complexity, special/edge case considerations etc)

File: src/marqo/tensor_search/tensor_search.py [hunk 3]
```diff
@@ -267,9 +267,53 @@ def _autofill_index_settings(index_settings: dict):
...
+            }
+        }
+    }
+
+    try:
+        doc_count = HttpRequests(config).post(path=F"{index_name}/_count")["count"]
+        vector_count = HttpRequests(config).get(path=f"{index_name}/_search", body=body) \
+            ["aggregations"]["nested_chunks"]["marqo_vector_count"]["value"]
+    except (KeyError, TypeError) as e:
+        error_message = (f"Marqo received an unexpected response from Marqo-os during execution of `get_stats()` APIs. "
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: We seem a bit inconsistent here, but `Marqo-OS` (caps) is better (both are used in the code base though). Also you can shorten this to "Marqo received an unexpected response from Marqo-OS. The expected fields do not exist in the response. Original error message = " (i.e. they already know they're calling get_stats API)

File: src/marqo/tensor_search/tensor_search.py [hunk 4]
```diff
@@ -267,9 +267,53 @@ def _autofill_index_settings(index_settings: dict):
...
+                         f"The expected fields do not exist in the response. Original error message = {e}")
+        logger.error(error_message)
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: Log and raise in an anti-pattern. Don't log when raising. ChatGPT explains it really well if you're wondering why

File: src/marqo/tensor_search/tensor_search.py [hunk 5]
```diff
@@ -267,9 +267,53 @@ def _autofill_index_settings(index_settings: dict):
...
+        raise errors.InternalError(error_message)
+    except (errors.IndexNotFoundError, errors.InvalidIndexNameError):
+        raise
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py:301]: This `except` seems superfluous. You're not doing anything with these error types.

File: src/marqo/tensor_search/tensor_search.py [hunk 6]
```diff
@@ -267,9 +267,53 @@ def _autofill_index_settings(index_settings: dict):
...
+    except errors.MarqoWebError as e:
+        error_message = (
+            f"Marqo encountered an error while communicating with Marqo-os during execution of `get_stats()` APIs. "
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: Same for Marqo-OS and "during execution of `get_stats() APIs" is unnecessary

Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: No need to log. See previous comment

File: tests/tensor_search/test_get_stats.py [hunk 7]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+                auto_refresh=True, device="cpu",
+                non_tensor_fields=["non_tensor_field"],
+                mappings={"my_multi_modal_field": {"type": "multimodal_combination", "weights": {
+                    "text_1": 0.5, "text_2": 0.8}}}
+            )
+        )
+
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfDocuments"] \
+               == expected_number_of_documents
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfVectors"] \
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py]: You're calling `tensor_search.get_stats` twice. You want to assert these against the response of a single API call.

File: tests/tensor_search/test_get_stats.py [hunk 8]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+            tensor_search.add_documents(
+                config=self.config, add_docs_params=AddDocsParams(
+                    index_name=self.index_name_1,
+                    auto_refresh=True, device="cpu",
+                    **test_case["add_docs_kwargs"],
+                )
+            )
+            assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfDocuments"] \
+                   == test_case["expected_number_of_documents"]
+            assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfVectors"] \
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py]: Same as above, best to call it once.

File: tests/tensor_search/test_get_stats.py [hunk 9]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+            config=self.config, add_docs_params=AddDocsParams(
+                index_name=self.index_name_1,
+                auto_refresh=True, device="cpu",
+                **test_case["add_docs_kwargs"],
+            )
+        )
+
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfDocuments"] \
+               == test_case["expected_number_of_documents"]
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfVectors"] \
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py]: Same as above, best to call it once.

File: tests/tensor_search/test_get_stats.py [hunk 10]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+                    "text_1": 0.5, "text_2": 0.8}}}
+            )
+        )
+
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfDocuments"] \
+               == expected_number_of_documents
+        assert tensor_search.get_stats(config=self.config, index_name=self.index_name_1)["numberOfVectors"] \
+               == expected_number_of_vectors
+
+    def test_get_stats_number_of_vectors_separated(self):
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py:71]: It would help to add a short description of what each test is trying to achieve for this one and the test before this.

File: tests/tensor_search/test_get_stats.py [hunk 11]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+
+    def test_IndexNotFoundError_error(self):
+        with patch("marqo.tensor_search.tensor_search.HttpRequests.get") as mock_get:
+            mock_get.side_effect = errors.IndexNotFoundError("IndexNotFoundError")
+
+            tensor_search.create_vector_index(config=self.config, index_name=self.index_name_1,)
+            with self.assertRaises(errors.IndexNotFoundError) as e:
+                res = tensor_search.get_stats(self.config, self.index_name_1)
+
+            self.assertIn("IndexNotFoundError", str(e.exception))
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py]: I think the previous assert covers this and is better too.

File: tests/tensor_search/test_get_stats.py [hunk 12]
```diff
@@ -14,25 +17,221 @@ def setUp(self) -> None:
...
+
+    def test_InvalidIndexNameError_error(self):
+        with patch("marqo.tensor_search.tensor_search.HttpRequests.get") as mock_get:
+            mock_get.side_effect = errors.InvalidIndexNameError("InvalidIndexNameError")
+
+            tensor_search.create_vector_index(config=self.config, index_name=self.index_name_1,)
+            with self.assertRaises(errors.InvalidIndexNameError) as e:
+                res = tensor_search.get_stats(self.config, self.index_name_1)
+
+            self.assertIn("InvalidIndexNameError", str(e.exception))
```
Comment (farshidz) [inline tests/tensor_search/test_get_stats.py]: The previous assert covers error type already

**Review by farshidz — CHANGES_REQUESTED** [review]

File: src/marqo/tensor_search/tensor_search.py (hunk 5 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py:301]: I do this because IndexNotFoundError and InvalidIndexNameError are inherited from MarqoWebError.  For these two errors, I want to raise it. For other errors, i want to convert them to internal errors.

Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py:301]: I updated the code to make it more straightforward

File: src/marqo/tensor_search/tensor_search.py [hunk 13]
```diff
@@ -298,18 +293,14 @@ def get_stats(config: Config, index_name: str):
...
-        raise
+        raise errors.InternalError(f"Marqo received an unexpected response from Marqo-OS. "
+                                   f"The expected fields do not exist in the response. Original error message = {e}")
     except errors.MarqoWebError as e:
-        error_message = (
-            f"Marqo encountered an error while communicating with Marqo-os during execution of `get_stats()` APIs. "
-            f"Original error message: {e.message}")
-        logger.error(error_message)
-        raise errors.InternalError(error_message)
+        if isinstance(e, (errors.IndexNotFoundError, errors.InvalidIndexNameError)):
```
Comment (farshidz) [inline src/marqo/tensor_search/tensor_search.py]: My bad about this one. Now that I know what you're doing the previous form was perfect

**Review by farshidz — CHANGES_REQUESTED** [review]

**Review by farshidz — APPROVED** [review]
